lnwallet: Introduce a fee buffer.

ziggie1984 opened the pull request:
Fixes #7657/#7721

This change introduces a fee buffer which accounts for a 100% fee rate increase plus an additional htlc output on the commitment transaction ~~at the current fee rate~~. This equals the behaviour what CLN does currently and is spec compliant IMO. It's a conservative approach because we are not taking any trimmed Htlcs at higher fee rates into account, but this has a positive side effect when fee rates decrease and dust htlcs become normal outputs (only relevant for non-anchor channels because they have to pay for the second-level covenant transaction).

~~Still need to write tests in the coming days. Change seems small at first glance but lets wait until tests are ready.~~




~~EDIT: Before updating all the unit tests and thinking of itests, I would be happy to receive first feedback on this approach (especially the additonal argument `enforceFeeBuffer` extra argument, maybe there is a cleaner approach?).~~

It is important to underline that the `fee buffer` is only a BOLT02 requirement, all the requirements regarding BOLT03 where the testcase go below the imaginable fee buffer still need hold.

This PR takes the recommendation from BOTL02 and calculates the fee buffer like the following:

`feeBuffer = currentFeeRate *2 * commitweight + additionalHTLCWeigth * currentFeeRate *2`

This is the same behaviour as eclair and cln follow. Subtle difference to eclair is that eclair accounts for the potential trimmed htlcs at higher fee rates whereas we do not account for this trimming (CLN does not as well). This is only relevant for non-anchor channels because anchor channels have zerofee htlc second-level covenant transactions

eclair: https://github.com/ACINQ/eclair/blob/97c9b579f319952c215f2a1930357de25a0640ab/eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala#L135-L151

cln: https://github.com/ElementsProject/lightning/blob/62ddf84b4f15a460ed5a8f72b313998a83eefe19/channeld/full_channel.c#L565-L569

So there are two occasions we have to differentiate when we want to enforce the fee buffer and when not.

1. When sending an payment or forwarding a payment we will enforce this fee buffer.
2. When evaluating a new FeeUpdate msg to increase the fee rate for a channel commitment we will not (only account for the additional HTLC a peer might add).


In addition what do you think of a new field when showing the channel: `CanSend`, which show the exact max amount one can send currently ?

alexbosworth commented on lnwallet/channel.go:
```diff
@@ -3841,7 +3841,25 @@ func (lc *LightningChannel) validateCommitmentSanity(theirLogCounter,
+		// (https://github.com/lightning/bolts/pull/867) introduced to
+		// the protocol, we should keep a fee buffer at least the size
+		// of twice the current feerate (BOLT 02). The reason for this
+		// is that updates form the peer and ourselves can happen
```
```suggestion
		// is that updates from the peer and ourselves can happen
```

Crypt-iQ commented on lnwallet/channel.go:
```diff
@@ -5875,7 +5947,7 @@ func (lc *LightningChannel) validateAddHtlc(pd *PaymentDescriptor) error {
 	err = lc.validateCommitmentSanity(
 		lc.remoteUpdateLog.logIndex, lc.localUpdateLog.logIndex,
-		false, pd, nil,
+		false, false, pd, nil,
```
rationale for `false` here?

ziggie1984 replied:
my bad, that was a mistake but would have been caught by the tests.

Roasbeef commented:
I wonder if we can write a fuzzer that it able to trigger the original edge case, which would then add confidence that this properly resolves the issue.

DerEwige commented:
@Roasbeef 

I encounter this edge case about once per week on my live node.
The reason why thi happens so much for me, is that I generate a lot of htlcs.
On a single channel it is usually about 1 htlc every 2-3s (if this channel is a source to me)

I stopped accepting channels from LND nodes that are sources to me, becaue if I accept those channels they end up in a FC usually in less than a week.

So if you wanna trigger this case.
Just empty the opner LND side untill you have only a few sats left and then spamm htlcs from both sides (and let them fail after a short time).
You should run into this edge case within a few minutes

ziggie1984 commented:
> I wonder if we can write a fuzzer that it able to trigger the original edge case, which would then add confidence that this properly resolves the issue.

I will definitely look into ways reproducing this error (fuzzing,itesting) and see whether this PR fixes the issues, good point.

Crypt-iQ commented:
> > I wonder if we can write a fuzzer that it able to trigger the original edge case, which would then add confidence that this properly resolves the issue.
> 
> I will definitely look into ways reproducing this error (fuzzing,itesting) and see whether this PR fixes the issues, good point.

It would probably be easiest to write a unit test in `channel_test.go` using these steps https://github.com/lightningnetwork/lnd/issues/7657#issuecomment-1696091228

ziggie1984 commented:
So I reworked this PR a bit from the initial version, so I generalized the part where we account for the `CommitmentFee` of the Channel introducing a new BufferType. So now we do have the feebuffer logic at one place only. Tho not super happy with the design therefore looking for some ideas of you guys how to make this think better. So looking for an `ApproachAck` so that I can make it ready for the final release.

I am actually not sure if we really need to FuzzTest this scenario, if so I would love to get some ideas how the pros here 😀.
I added the relevant unit tests which I think follow the case which eugene pointed out in his comment. 

I left some TODO's in this PR which I want you to focus on while reviewing, especially the proposal in https://github.com/lightningnetwork/lnd/issues/8249 where we would allow the peer to put us below our reserve if we are the opener of the channel.

So this PR is not final yet but ready to receive a Design Review.

ziggie1984 commented on lnwallet/channel.go:
```diff
@@ -3896,16 +4035,26 @@ func (lc *LightningChannel) validateCommitmentSanity(theirLogCounter,
+	commitFeeMsat := lnwire.NewMSatFromSatoshis(commitFee)
+
 	switch {
+	// TODO(ziggie): Remove the InitialBalance check?
```
So I thought about the argument in https://github.com/lightningnetwork/lnd/issues/8249  and I think we should include this if we wanna fix the stuck channel and a potential splicing case where the reserve can increase and we need to push funds to the remote. I did not include it in a commit for now because it broke several testcases again, so first waiting on your comments.
```

	ourBalanceDecreasing := ourBalance < ourInitialBalance
	theirBalanceDecreasing := theirBalance < theirInitialBalance

	switch {
	// We do not allow the dipping below the channel reserve if their
	// balance is NOT decreasing which means they are not adding an htlc.
	case !theirBalanceDecreasing && ourBalance < ourReserve:
		lc.log.Debugf("Funds below chan reserve: ourBalance=%v, "+
			"ourReserve=%v, commitFee=%v, feeBuffer=%v "+
			"chan_initiator=%v", ourBalance, ourReserve,
			commitFeeMsat, bufferAmt, lc.channelState.IsInitiator)

		return fmt.Errorf("%w: our balance below chan reserve",
			ErrBelowChanReserve)

	// We do not allow the dipping below their channel reserve if our amount
	// is NOT decreasing which means we are not adding any htlcs.
	case !ourBalanceDecreasing && theirBalance < theirReserve:
		lc.log.Debugf("Funds below chan reserve: theirBalance=%v, "+
			"theirReserve=%v", theirBalance, theirReserve)

		return fmt.Errorf("%w: their balance below chan reserve",
			ErrBelowChanReserve)

	// We allow our balance to dip below the channel reserve in case the
	// balance of our peer is decreasing which means an incoming htlc is
	// added to the commitment state. We carve out this exception to account
	// for stuck channel prior to introducing the fee buffer and in forsight
	// of splicing where increasing the channel capacity can increase the
	// reserve requirement.
	case theirBalanceDecreasing && ourBalance < ourReserve:
		lc.log.Debugf("Funds below chan reserve but our amount is "+
			"increasing: ourInitialBalance=%v, ourBalance=%v, "+
			"ourReserve=%v", ourInitialBalance, ourBalance,
			ourReserve)

	// We allow their balance to dip below the channel reserve in case our
	// balance is decreasing which means we are adding an outgoing htlc to
	// the commitment state. We carve out this exception to account
	// for stuck channel prior to introducing the fee buffer and in forsight
	// of splicing where increasing the channel capacity can increase the
	// reserve requirement.
	case ourBalanceDecreasing && theirBalance < theirReserve:
		lc.log.Debugf("Funds below chan reserve but their amount is "+
			"increasing: theirInitialBalance=%v, ourBalance=%v, "+
			"ourReserve=%v", theirInitialBalance, ourBalance,
			ourReserve)
	}
```

ziggie1984 replied:
agree, will reformulate the TODO a bit, because the `ourInitialBalance ` will come useful so no need to remove it.

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@crypt-iq: review reminder
@ziggie1984, remember to re-request review from reviewers when ready

TrezorHannes commented:
Can we please please go ahead and get this one reviewed and prioritised for release? This is the 3rd FC I had to compensate for, 2 with LND peers and 1 with CLN. 

Started 7hrs ago with the insufficient bandwidth, and then FC'ed by CLN for 100k sats
`23c251a87f9b39b47c412708fef703388f0ba4bac4e6bf638fd4f60063aa1cf4`

```terminal
2023-12-18 13:22:29.815 [WRN] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): insufficient bandwidth to route htlc: 99130847 mSAT is larger than 60606786 mSAT
2023-12-18 13:22:41.921 [WRN] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): insufficient bandwidth to route htlc: 150653656 mSAT is larger than 4996303 mSAT
2023-12-18 13:23:06.687 [WRN] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): insufficient bandwidth to route htlc: 97009797 mSAT is larger than 81261177 mSAT
2023-12-18 13:23:11.180 [WRN] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): insufficient bandwidth to route htlc: 99126288 mSAT is larger than 81261177 mSAT
2023-12-18 13:23:46.791 [WRN] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): insufficient bandwidth to route htlc: 75326828 mSAT is larger than 63491678 mSAT
2023-12-18 13:24:44.154 [ERR] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): failing link: unable to update commitment: commitment transaction dips peer below chan reserve: our balance below chan reserve with error: internal error
2023-12-18 13:24:44.154 [ERR] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): link failed, exiting htlcManager
2023-12-18 13:24:44.154 [INF] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): exited
2023-12-18 13:24:44.154 [INF] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): stopping
2023-12-18 13:24:45.539 [INF] PEER: Peer(02d96eadea3d780104449aca5c93461ce67c1564e2e1d73225fa67dd3b997a6018): Loading ChannelPoint(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1), isPending=false
2023-12-18 13:24:45.539 [INF] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): starting
2023-12-18 13:24:45.539 [INF] CNCT: Attempting to update ContractSignals for ChannelPoint(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1)
2023-12-18 13:24:45.539 [INF] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): HTLC manager started, bandwidth=0 mSAT
2023-12-18 13:24:45.539 [INF] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): Attempting to re-synchronize channel: SCID=806597:1249:1, status=ChanStatusDefault, initiator=true, pending=false, local commitment has height=1282587, local_htlc_index=561647, local_log_index=663648, remote_htlc_index=101246, remote_log_index=662891, remote commitment has height=1287392, local_htlc_index=561648, local_log_index=663649, remote_htlc_index=101245, remote_log_index=662890
2023-12-18 13:24:45.835 [INF] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): received re-establishment message from remote side
2023-12-18 13:24:46.414 [WRN] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): insufficient bandwidth to route htlc: 37663413 mSAT is larger than 0 mSAT
2023-12-18 13:24:46.775 [WRN] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): insufficient bandwidth to route htlc: 10004320 mSAT is larger than 0 mSAT
2023-12-18 13:24:51.645 [ERR] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): failing link: ChannelPoint(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): unable to accept new commitment: commitment transaction dips peer below chan reserve: our balance below chan reserve with error: invalid commitment
2023-12-18 13:24:51.645 [ERR] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): link failed, exiting htlcManager
2023-12-18 13:24:51.645 [INF] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): exited
2023-12-18 13:24:51.645 [INF] HSWC: ChannelLink(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): stopping
2023-12-18 13:24:51.645 [INF] CNCT: Attempting to force close ChannelPoint(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1)
2023-12-18 13:24:51.646 [INF] NANN: Announcing channel(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1) disabled [requested]
2023-12-18 13:24:51.674 [INF] CNCT: ChannelArbitrator(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): force closing chan
2023-12-18 13:24:51.700 [INF] CNCT: Broadcasting force close transaction b56fd38624428f190b03875c30518b21a35c3f8267c9d60554f05c17a4d3b9f0, ChannelPoint(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): (*wire.MsgTx)(0xc019a05c00)({
   PreviousOutPoint: (wire.OutPoint) 41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1,
2023-12-18 13:24:51.762 [INF] CNCT: ChannelArbitrator(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): trigger userTrigger moving from state StateCommitmentBroadcasted to StateCommitmentBroadcasted
2023-12-18 13:43:20.088 [INF] NTFN: Dispatching confirmed spend notification for outpoint=41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1, script=0 3c1e049899c6129a13841d790a78fe7dea076ba072533e6bf665aa39cc70b84e at current height=821769: 23c251a87f9b39b47c412708fef703388f0ba4bac4e6bf638fd4f60063aa1cf4[0] spending 41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1 at height=821769
2023-12-18 13:43:20.091 [INF] CNCT: Unilateral close of ChannelPoint(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1) detected
2023-12-18 13:43:20.093 [INF] CNCT: ChannelArbitrator(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): remote party has closed channel out on-chain
2023-12-18 13:43:25.663 [INF] CNCT: ChannelArbitrator(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): trigger remoteCloseTrigger moving from state StateCommitmentBroadcasted to StateContractClosed
2023-12-18 13:43:25.947 [INF] CNCT: ChannelArbitrator(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): still awaiting contract resolution
2023-12-18 13:43:26.143 [INF] CNCT: *contractcourt.commitSweepResolver(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): Sweeping with witness type: CommitmentToRemoteConfirmed
2023-12-18 13:43:26.143 [INF] CNCT: *contractcourt.commitSweepResolver(41cd011365aea32e9089ba9d927a356b43bdcb11d8802d725e0df523e14455e1:1): sweeping commit output
```

KnockOnWoodNode commented:
Count me in as another angry Karen.

```
[INF] PEER: Peer(035e4ff418fc8b5554c5d9eea66396c227bd429a3251c8cbc711002ba215bfc226): Loading ChannelPoint(65fc3165c20777c23e4ace1ead4140d9adc3a0a742339682747e2db14cfe1fe4:4), isPending=false
[INF] HSWC: ChannelLink(65fc3165c20777c23e4ace1ead4140d9adc3a0a742339682747e2db14cfe1fe4:4): starting
[INF] HSWC: ChannelLink(65fc3165c20777c23e4ace1ead4140d9adc3a0a742339682747e2db14cfe1fe4:4): HTLC manager started, bandwidth=0 mSAT
[INF] CNCT: Attempting to update ContractSignals for ChannelPoint(65fc3165c20777c23e4ace1ead4140d9adc3a0a742339682747e2db14cfe1fe4:4)
[INF] HSWC: ChannelLink(65fc3165c20777c23e4ace1ead4140d9adc3a0a742339682747e2db14cfe1fe4:4): Attempting to re-synchronize channel: SCID=799216:842:4, status=ChanStatusDefault, initiator=true, pending=false, local commitment has height=675964, local_htlc_index=263955, local_log_index=347549, remote_htlc_index=82473, remote_log_index=346428, remote commitment has height=673410, local_htlc_index=263956, local_log_index=347550, remote_htlc_index=82471, remote_log_index=346426
[INF] HSWC: ChannelLink(65fc3165c20777c23e4ace1ead4140d9adc3a0a742339682747e2db14cfe1fe4:4): received re-establishment message from remote side
[ERR] HSWC: ChannelLink(65fc3165c20777c23e4ace1ead4140d9adc3a0a742339682747e2db14cfe1fe4:4): failing link: unable to update commitment: commitment transaction dips peer below chan reserve: our balance below chan reserve with error: internal error
[ERR] HSWC: ChannelLink(65fc3165c20777c23e4ace1ead4140d9adc3a0a742339682747e2db14cfe1fe4:4): link failed, exiting htlcManager
[INF] HSWC: ChannelLink(65fc3165c20777c23e4ace1ead4140d9adc3a0a742339682747e2db14cfe1fe4:4): stopping
[INF] HSWC: ChannelLink(65fc3165c20777c23e4ace1ead4140d9adc3a0a742339682747e2db14cfe1fe4:4): exited
```

A FC with WOS is incoming in 44hrs if fees do not go down sensibly in the meantime, and I do not expect to get out of this because this is released in the nick of time... still, angry Karen,

kilrau commented:
Boltz LND just had at least one force close because of this too. It wasn't cheap. We actually only realized because it was related to a failed swap. Might have had more in the past.

We'd highly appreciate giving this some priority! :pray: @Roasbeef

DerEwige commented:
My most recent force close because of this issue on an eclair node:



```
2023-12-18 22:14:42,193 WARN  f.a.e.c.fsm.Channel n:03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd c:e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 - remote can't afford increased commit tx fees once new HTLC is added: missing=1381 sat reserve=20000 sat fees=13044 sat while processing cmd=CMD_ADD_HTLC in state=NORMAL
2023-12-18 22:14:42,968 WARN  f.a.e.c.fsm.Channel n:03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd c:e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 - remote can't afford increased commit tx fees once new HTLC is added: missing=1381 sat reserve=20000 sat fees=13044 sat while processing cmd=CMD_ADD_HTLC in state=NORMAL
2023-12-18 22:14:54,672 DEBUG f.a.e.p.r.ChannelRelay PAY h:6cc8e968405df3e334ab823a996c225b9e8a91b11dc0587f8606e24ec9b68538 p:f69047f2 - candidate channel: channelId=e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 availableForSend=1727717443 msat capacity=2000000 sat channelUpdate=ChannelUpdate(6492f395518311bd7972fb81824b1f4a118312c504f4c7ef760ab022da0a5aba4f87833c3a097ba385e7dbedd392ee360ff5464a94fc7b3bb314f0211f3240cd,6fe28c0ab6f1b372c1a6a246ae63f74f931e8365e15a089c68d6190000000000,820620x2237x1,1702933742 unixsec,MessageFlags(false),ChannelFlags(true,true),CltvExpiryDelta(144),1 msat,1 msat,43,1024000000 msat,TlvStream(Set(),Set())) result=success
2023-12-18 22:14:54,672 INFO  f.a.e.p.r.ChannelRelay PAY h:6cc8e968405df3e334ab823a996c225b9e8a91b11dc0587f8606e24ec9b68538 p:f69047f2 - forwarding htlc to channelId=e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31
2023-12-18 22:14:54,672 WARN  f.a.e.c.fsm.Channel n:03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd c:e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 - remote can't afford increased commit tx fees once new HTLC is added: missing=1381 sat reserve=20000 sat fees=13044 sat while processing cmd=CMD_ADD_HTLC in state=NORMAL
2023-12-18 22:15:06,933 DEBUG f.a.e.p.r.ChannelRelay PAY h:d60b936d815d00eef3cbddb76de3899e1585cba29396a9a670ee631d1f59714d p:e1ad9114 - candidate channel: channelId=e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 availableForSend=1727717443 msat capacity=2000000 sat channelUpdate=ChannelUpdate(6492f395518311bd7972fb81824b1f4a118312c504f4c7ef760ab022da0a5aba4f87833c3a097ba385e7dbedd392ee360ff5464a94fc7b3bb314f0211f3240cd,6fe28c0ab6f1b372c1a6a246ae63f74f931e8365e15a089c68d6190000000000,820620x2237x1,1702933742 unixsec,MessageFlags(false),ChannelFlags(true,true),CltvExpiryDelta(144),1 msat,1 msat,43,1024000000 msat,TlvStream(Set(),Set())) result=success
2023-12-18 22:15:06,933 INFO  f.a.e.p.r.ChannelRelay PAY h:d60b936d815d00eef3cbddb76de3899e1585cba29396a9a670ee631d1f59714d p:e1ad9114 - forwarding htlc to channelId=e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31
2023-12-18 22:15:06,933 WARN  f.a.e.c.fsm.Channel n:03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd c:e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 - remote can't afford increased commit tx fees once new HTLC is added: missing=1381 sat reserve=20000 sat fees=13044 sat while processing cmd=CMD_ADD_HTLC in state=NORMAL
2023-12-18 22:15:20,259 WARN  f.a.e.c.fsm.Channel n:03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd c:e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 - waited for too long for a revocation to remoteCommitNumber=142030, disconnecting
2023-12-18 22:15:20,259 WARN  f.a.e.c.fsm.Channel n:03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd c:e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 - unhandled event peer 03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd disconnecting in state NORMAL
2023-12-18 22:15:22,155 INFO  f.a.e.i.Peer CON n:03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd c:e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 - OUT msg=ChannelReestablish(e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31,140333,142030,<private_key>,0373cf0fa82c0bb976b00c910403597aea5f5e4b7f7d2df1f1255e5d04286a019e,TlvStream(Set(),Set()))
2023-12-18 22:15:22,706 INFO  f.a.e.i.Peer CON n:03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd c:e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 - IN msg=ChannelReestablish(e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31,142031,140332,<private_key>,030cc80216707086103f814cd5c6a5c6f41d6b3d3d61d5b7577383a3e9d3495e79,TlvStream(Set(),Set()))
2023-12-18 22:15:22,706 INFO  f.a.e.i.Peer n:03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd c:e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 - OUT msg=CommitSig(e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31,0b282f4a09f53d4f6676f7ba4afc9dd28d34660d5d263620ae3a91ed30fabd1409f1112e98888c9fdca0bb9002666ef8ba30af0e0a5f17f6b830ea42b1669a8d,List(369fa9264302e2755c5808e4e4a43643f8718e708015d6a039097777dfd817ac770a2753665097b4d4c05ddee5c75690bd4e1004b8131ca194700dba4f034543, 24c82dc486db818e987899cf12ba53dd6711b79da6ffcc117ecea39a5a0f38ab37dda36697c4e7426c4cc1e37b757fc339740d86f33576e53ab6c4e25e1b3b17, 4ae07531c31e0021a7fb0889278e8afe93d2dfdbd035877eb36e07cf1536066758cc2da9a1ab0730266e74a9b93e48fe542040e24611729d69a69338d4e29287, 77bb24419d4ee84feaa03e7f04dd64cc247c0b033194f288fec331dedad4a3ef790a80ecd9dfda73d6b47083d0137b70f50715714abfb4313f86abc774a95b20, fede83ca8eda4673f5c13cbfcc90a9150767d60e4e0061abd6f6b581ba2cce8c289b783c4c3f8b3516a880f4aa7abb8103c43a777bf9f7ddd705abe033b6c63a),TlvStream(Set(),Set()))
2023-12-18 22:15:42,456 INFO  f.a.e.i.Peer CON n:03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd c:e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 - OUT msg=ChannelReestablish(e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31,140333,142030,<private_key>,0373cf0fa82c0bb976b00c910403597aea5f5e4b7f7d2df1f1255e5d04286a019e,TlvStream(Set(),Set()))
2023-12-18 22:15:42,652 INFO  f.a.e.i.Peer CON n:03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd c:e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 - IN msg=ChannelReestablish(e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31,142031,140332,<private_key>,030cc80216707086103f814cd5c6a5c6f41d6b3d3d61d5b7577383a3e9d3495e79,TlvStream(Set(),Set()))
2023-12-18 22:15:42,653 INFO  f.a.e.i.Peer n:03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd c:e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 - OUT msg=CommitSig(e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31,0b282f4a09f53d4f6676f7ba4afc9dd28d34660d5d263620ae3a91ed30fabd1409f1112e98888c9fdca0bb9002666ef8ba30af0e0a5f17f6b830ea42b1669a8d,List(369fa9264302e2755c5808e4e4a43643f8718e708015d6a039097777dfd817ac770a2753665097b4d4c05ddee5c75690bd4e1004b8131ca194700dba4f034543, 24c82dc486db818e987899cf12ba53dd6711b79da6ffcc117ecea39a5a0f38ab37dda36697c4e7426c4cc1e37b757fc339740d86f33576e53ab6c4e25e1b3b17, 4ae07531c31e0021a7fb0889278e8afe93d2dfdbd035877eb36e07cf1536066758cc2da9a1ab0730266e74a9b93e48fe542040e24611729d69a69338d4e29287, 77bb24419d4ee84feaa03e7f04dd64cc247c0b033194f288fec331dedad4a3ef790a80ecd9dfda73d6b47083d0137b70f50715714abfb4313f86abc774a95b20, fede83ca8eda4673f5c13cbfcc90a9150767d60e4e0061abd6f6b581ba2cce8c289b783c4c3f8b3516a880f4aa7abb8103c43a777bf9f7ddd705abe033b6c63a),TlvStream(Set(),Set()))
2023-12-18 22:15:47,737 INFO  f.a.e.i.Peer n:03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd c:e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 - IN msg=Error(e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31,ByteVector(18 bytes, 0x696e76616c696420636f6d6d69746d656e74),TlvStream(Set(),Set()))
2023-12-18 22:15:47,737 ERROR f.a.e.c.fsm.Channel n:03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd c:e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 - peer sent error: ascii='invalid commitment' bin=696e76616c696420636f6d6d69746d656e74
2023-12-18 22:15:47,739 ERROR f.a.e.c.fsm.Channel n:03c157946cc1cd376b929e36006e645fae490b1b1d4156b40db804e01b4bda48cd c:e94b2f891828815cad73db558483e32aea71fdcd1df48d38980ce0ec8802ed31 - force-closing with fundingIndex=0

```

yyforyongyu commented on lnwallet/channel.go:
```diff
@@ -5736,13 +5920,19 @@ func (lc *LightningChannel) InitNextRevocation(revKey *btcec.PublicKey) error {
 
 	pd := lc.htlcAddDescriptor(htlc, openKey)
-	if err := lc.validateAddHtlc(pd); err != nil {
+	if err := lc.validateAddHtlc(pd, opts.buffer); err != nil {
```
looks like `AddHTLC` will only be called with `FeeBuffer` option, so I guess the `AddHtlcOptions` can be removed to simply the code?

ziggie1984 replied:
The reason I introduced the variadic parameter is, if I would add a normal function parameter for AddHTLC, I would need to edit all the tests where `AddHTLC` is used (a lot) and would need to add a function parameter although none of them use a buffer anyways. But also happy to do the work and change if you think that's a cleaner approach.

ziggie1984 replied:
I agree, I think we should get rid of the variadic param and do the work and make sure all testcases which are not BOLT03 testcases adhere to the buffer.

ziggie1984 replied:
I came up with a slightly different approach, I enhanced the method with an unexported method `addHTLC` which is needed to test all the scenarios e.g. BOLT03 test vectors and kept the `AddHTLC` function signature and always enforcing the feebufffer in it, lmkwyt ?

yyforyongyu commented on lnwallet/channel.go:
```diff
@@ -3818,14 +3818,146 @@ func (lc *LightningChannel) getUnsignedAckedUpdates() []channeldb.LogUpdate {
+// balance and the exact buffer amount (excluding the commitment fee).
+func (lc *LightningChannel) applyCommitFee(
+	balance lnwire.MilliSatoshi, commitWeight int64,
+	feePerKw chainfee.SatPerKWeight, isLocal bool,
```
so this `isLocal` is only used for logging?

ziggie1984 replied:
Yes removed it and tried a different way.

yyforyongyu commented on lnwallet/channel.go:
```diff
@@ -3854,39 +3986,46 @@ func (lc *LightningChannel) validateCommitmentSanity(theirLogCounter,
+			return err
+		}
+	} else {
+		theirBalance, bufferAmt, err = lc.applyCommitFee(
```
So if we are not the channel funder we won't check their balance to account for the fee buffer? Why dont we also check it?

ziggie1984 replied:
That's a very good idea 👌, this will help to prevent this bug when only one side has been updated to the newest lnd release, nice !

ziggie1984 replied:
Thought about this a second time and I think we cannot do this. When we also enforce a feebuffer when sending but the other node does not want to enforce the feebuffer (still on older software) we can end up in a scenario that the peer depletes his side even more and now the channel is stuck because he cannot rebalance into the channel because we will not put any HTLC on it. So we should not intervene when we are NOT the opener I think.

yyforyongyu reviewed (commented):
Very important work! Concept ACK. Some notes may be helpful for other reviewers, the core change happens in `applyCommitFee`, which is called by,

```
`validateCommitmentSanity`, and used in various methods:
	- `SignNextCommitment` - `NoBuffer`
	- `ReceiveNewCommitment` - `NoBuffer`
	- `ReceiveHTLC` - `NoBuffer`
	- `validateAddHtlc`
		- `AddHTLC` - options - `FeeBuffer`
		- `MayAddOutgoingHtlc` - `FeeBuffer`

`availableCommitmentBalance`, and used in various methods:
	- `availableBalance`
		- `validateFeeRate` - `AdditionalHtlc`
			- `UpdateFee` which sends `update_fee` msg
		- `AvailableBalance` - `FeeBuffer` - used in link to decide channel bandwith
		- `MaxFeeRate` - `AdditionalHtlc` TODO(yy): downgrade it to private method
			- `IdealCommitFeeRate` - used in link to decide new commit fee rate
```

To summarize, when we add a htlc, we enforce a fee buffer. When we receive an HTLC or `commit_sig` or sign a new `commit_sig`, we don't check for fee buffer. When we send a `update_fee` message, we only account one more htlc as our fee buffer.

My questions and suggestions,
- not sure about `AdditionalHtlc`'s use case. I think it makes sense to only account for one extra HTLC when we know we will do a coop close, to make sure there's enough budget to pay the final tx fees. Not sure tho, if there another round of commitment dance, then this budget may not be enough? I guess in that case we can always fail it if it cannot meet the fee buffer?

- maybe we don't change `validateCommitmentSanity`, use `FeeBuffer` and add code/method to validate fee buffer in `validateAddHtlc` directly, seems to give us less code change.

ziggie1984 commented on lnwallet/channel.go:
```diff
@@ -8441,8 +8663,9 @@ func (lc *LightningChannel) MaxFeeRate(maxAllocation float64) chainfee.SatPerKWe
-	availableBalance, weight := lc.availableBalance()
+	// balance. We do not account for a fee buffer here because that is
+	// exactly why it was introduced to react for sharp fee changes.
+	availableBalance, weight := lc.availableBalance(AdditionalHtlc)
```
Hmm before laying out why I used `AdditionalHtlc` I try to explain why I don't used the other two in this case:

Why `FeeBuffer` was NOT used:

We are talking about updating the fee of the commitment here, that's one of the reasons the feebuffer was introduced which we keep when sending HTLCs? It's also referred as `fee spike buffer` in the Bolts. So when we would also use the buffer here we would not really be using it when feespikes happen ? Sure this means when we are drained locally and we update the fee, we might not be able to send further htlcs in the next commitment dance but because we use `AdditionalHtlc` we can still rebalance this HTLC and fill up local again
 
```
SHOULD NOT offer amount_msat if, after adding that HTLC to its commitment transaction, its remaining balance doesn't allow it to pay the commitment transaction fee when receiving or sending a future additional non-dust HTLC while maintaining its channel reserve. It is recommended that this "fee spike buffer" can handle twice the current feerate_per_kw to ensure predictability between implementations.
``` 

Why `NoBuffer` was NOT used:

Especially the edge case where we use 1 for `maxAllocation` and don't keep space for another HTLC, we might make the Channel unusable (only dust-htlcs might be sent) because no party can add an HTLC to the channel.

Thats why I decided to introduce this new Buffer type: `AdditionalHtlc `

But I might misunderstood something here.

yyforyongyu replied:
> So when we would also use the buffer here we would not really be using it when feespikes happen ? 

Exactly, `update_fee` is the place to use the fee buffer.

> Sure this means when we are drained locally and we update the fee, we might not be able to send further htlcs in the next commitment dance but because we use AdditionalHtlc we can still rebalance this HTLC and fill up local again

I'm not sure about this. Suppose we wanna rebalance, another round of `update_add_htlc` is sent and received. However, even this one htlc cannot be added because we require `FeeBuffer` when adding and receiving HTLCs?

ziggie1984 replied:
> I'm not sure about this. Suppose we wanna rebalance, another round of update_add_htlc is sent and received. However, even this one htlc cannot be added because we require FeeBuffer when adding and receiving HTLCs?

We do not enforce the buffer when receiving an htlc, because we are not in control as soon as the Peer adds an HTLC to our Commitment we can only accept it, otherwise the channel is borked.

> Exactly, update_fee is the place to use the fee buffer.

Do you think `AdditionalHTLC` is the right buffer type here then, I am still not sure if our understanding is the same 😅 ?

ziggie1984 commented:
> maybe we don't change validateCommitmentSanity, use FeeBuffer and add code/method to validate fee buffer in validateAddHtlc directly, seems to give us less code change.

I think this might be a good idea, will try to code this variant and see if I can make it more clean 👍

> not sure about AdditionalHtlc's use case. I think it makes sense to only account for one extra HTLC when we know we will do a coop close, to make sure there's enough budget to pay the final tx fees. Not sure tho, if there another round of commitment dance, then this budget may not be enough? I guess in that case we can always fail it if it cannot meet the fee buffer?


hmm I think the `AdditionalHtlc` makes sense when we handle the calculation of the MaxFeeRate, because the `maxAllocation ` can be 1 and if fees spiked extremely we could end up in an unusable channel (only dust would be sendable)? I haven't thought about keeping an extra buffer for the coop close, but I think the coop close should not negotiate fees which are higher than the current commitment feerate, because that would reveal that the commitment feerate is too low and should be updated in the first place? But maybe I am missing something here.

ziggie1984 commented on lnwallet/channel.go:
```diff
@@ -3818,14 +3818,146 @@ func (lc *LightningChannel) getUnsignedAckedUpdates() []channeldb.LogUpdate {
+	NoBuffer BufferType = iota
+
+	// FeeBuffer accounts for several edge cases. One of them is where
+	// a locally drained channel might become unusable due to the non-opener
```
Hmm I really meant unusable = stuck ?

ziggie1984 commented on lnwallet/channel.go:
```diff
@@ -3818,14 +3818,146 @@ func (lc *LightningChannel) getUnsignedAckedUpdates() []channeldb.LogUpdate {
+		}
+	}
+
+	lc.log.Debugf("Balance(is_local=%b) is negative: balance=%v, "+
```
Voted for error.

ziggie1984 commented on lnwallet/channel.go:
```diff
@@ -8130,6 +8347,9 @@ func (lc *LightningChannel) availableCommitmentBalance(view *htlcView,
 	nonDustHtlcAmt := dustlimit + htlcFee
+	// CommitFee when the peer is the initiator of the channel (without
+	// the fee buffer).
+	htlcCommitFee := commitFee + additionalHtlcFee
```
hmm it is used in the followed if clause where we potentially decrease our available limit to the `dustlimit -1` when the peer cannot pay an additional htlc. Changed the name of the variable to: `commitFeeWithHtlc`

ziggie1984 commented on lnwallet/channel.go:
```diff
@@ -8070,26 +8303,45 @@ func (lc *LightningChannel) availableCommitmentBalance(view *htlcView,
+		ourBalance, bufferAmt, err = lc.applyCommitFee(
+			ourBalance, futureCommitWeight, feePerKw, buffer,
+		)
+		// TODO(ziggie): Should we add complexity here and return the
```
what do you think about this @yyforyongyu  ?

ziggie1984 replied:
What about keeping this at 0, but adding a `CanSend` field for the Channel (OpenChannel I think), why bother here in returning a dust amount, if we cannot add an HTLC, we should signal 0, so that we are not draining the channel further.

yyforyongyu reviewed (changes requested):
Looking good. Still need to do another round to review the tests. Meanwhile sharing some notes, I think it's helpful to unify the concepts that are used here re the value stored in a channel,
```
┌─────────────────────┐
│ ┌─────────────────┐ │
│ │                 │ │
│ │     routing     │ │
│ │                 │ │
│ │    bandwidth    │ │
│ │                 │ │
│ ├─────────────────┤ │
│ │    fee buffer   │ │
│ ├─────────────────┤ │
│ │ channel reserve │ │
│ └─────────────────┘ │
│                     │
│   channel balance   │
│                     │
├─────────────────────┤
│     commit fees     │
└─────────────────────┘
```

`commit fees` is the onchain fee, and `channel balance` is what the user will receive once the channel is closed, which is made of three parts,
- `channel reserve`, the security deposit.
- `fee buffer`, a dedicated amount that can be used to deal with fee spikes, and we expect `commit fees <= fee buffer`.
- `routing bandwidth`, the value that can be used for routing, ie, add htlcs.

If we can agree on these concepts we could make future PRs to apply them in our code, which can make things easier.

In a following PR I think we should make the fee buffer configurable, something like `currentFeeRate * BufferFactor * (commitweight + additionalHTLCWeigth * NumBufferedHTLCs)`

ziggie1984 commented on lnwallet/channel.go:
```diff
@@ -3896,16 +4041,32 @@ func (lc *LightningChannel) validateCommitmentSanity(theirLogCounter,
+	commitFeeMsat := lnwire.NewMSatFromSatoshis(commitFee)
+
 	switch {
+	// TODO(ziggie): Allow the peer dip us below the channel reserve when
```
I think you are right we definitly need to update the spec here to make sure we force-close the channel if this happens instead of just sending out a warning, because the channel is done. But I am not sure whether we should accept the HTLC which implies sending the `update_fail_htlc `, because we don't want to advance the channelstate if it dips us below the reserve. Maybe we need to special case it when we introduce the exception for splicing will talk about it in the BOLT Spec Meeting.

yyforyongyu replied:
A second thought is, it's OK to send `warning` or `error` to stop the link - on our end, we'd make sure it won't happen, and if they make it happen, then it's somewhat reasonable to think the peer is violating the protocol so we can terminate - though in reality we know it's better to loosen it a bit.

On the other hand, I think we should definitely distinguish the case when the channel drops below fee buffer vs drops below channel reserve. We should never fail the link when it drops below fee buffer, since that's what the buffer is used for - to buffer the concurrent add updates that may happen to a given channel. So imo we should somehow cancel that htlc if it will make the channel drop below fee buffer.

Re `update_fail_htlc`, that's a very good point. I think if we are the sender of the htlc, we can easily cancel it back to the upper subsystems instead of sending an `error` to the peer. And if we are the receiver of the htlc, since we don't enforce the fee buffer check, we cannot do anything. But this sort makes me wondering we should also check the fee buffer for the remote balance, and sends back a `warning` message.

In addition, we should treat `error` and `warning` messages differently, need to check how we handle them more.

ziggie1984 replied:
> On the other hand, I think we should definitely distinguish the case when the channel drops below fee buffer vs drops below channel reserve. We should never fail the link when it drops below fee buffer, since that's what the buffer is used for - to buffer the concurrent add updates that may happen to a given channel. So imo we should somehow cancel that htlc if it will make the channel drop below fee buffer.

Agree currently we make sure that we only enforce the buffer when we are in control of also aborting the htlc process, so I think we already do it but its currently not set into stone meaning that the `validateAddHtlc` can be called with any buffer in the future so maybe we should introduce extra functions because of the safety maybe ?

ziggie1984 commented on lnwallet/channel.go:
```diff
@@ -8070,26 +8283,45 @@ func (lc *LightningChannel) availableCommitmentBalance(view *htlcView,
+		// tx.
+		futureCommitWeight := commitWeight
+		if buffer == FeeBuffer {
+			futureCommitWeight += input.HTLCWeight
```
Those 2 are different, the one we are adding here, is the HTLC we would use in sending the HTLC we are evaluating here, the other HTLC is for the remote peer.

ziggie1984 commented on lnwallet/channel.go:
```diff
@@ -8130,18 +8362,21 @@ func (lc *LightningChannel) availableCommitmentBalance(view *htlcView,
 	// The HTLC output will be manifested on the commitment if it
 	// is non-dust after paying the HTLC fee.
 	nonDustHtlcAmt := dustlimit + htlcFee
+	// CommitFee when the peer is the initiator of the channel (without
```
yes changed the comment.

ziggie1984 commented on lnwallet/channel.go:
```diff
@@ -3818,14 +3818,144 @@ func (lc *LightningChannel) getUnsignedAckedUpdates() []channeldb.LogUpdate {
+// applyCommitFee applies the commitFee including a buffer to the balance amount
+// and verifies that it does not become negative. This function returns the new
+// balance and the exact buffer amount (excluding the commitment fee).
+func (lc *LightningChannel) applyCommitFee(
```
Added a unittest. I am also checking in this function whether we are the initiator before allowing to apply the FeeBuffer. So I think it makes sense to keep this method part of the `*LightningChannel` pointer receiver ?

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -1656,7 +1656,9 @@ func TestChannelBalanceDustLimit(t *testing.T) {
 
 	htlc, preimage := createHTLC(0, htlcAmount)
-	aliceHtlcIndex, err := aliceChannel.AddHTLC(htlc, nil)
+	// We need to use `addHTLC` instead of `AddHTLC` so that we do not
```
nit: newline

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -4826,12 +4827,22 @@ func TestChanAvailableBandwidth(t *testing.T) {
+		commitFee := lnwire.NewMSatFromSatoshis(aliceState.CommitFee)
+		commitWeight := CommitWeight(cType)
+		commitWeight += numNonDustHtlcsOnCommit * input.HTLCWeight
+		// Add weight for an additional htlc because this is also done
```
nit: newline

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -4826,12 +4827,22 @@ func TestChanAvailableBandwidth(t *testing.T) {
 		// current local commitment balance, minus the channel reserve
 		// and the fee for adding an HTLC.
-		expBalance := aliceBalance - aliceReserve - htlcFee
+		expBalance := aliceBalance + commitFee - aliceReserve -
```
nice👍

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -4977,7 +4996,11 @@ func TestChanAvailableBalanceNearHtlcFee(t *testing.T) {
 
 		htlc, preImage := createHTLC(int(htlcIndex), htlcAmt)
-		if _, err := aliceChannel.AddHTLC(htlc, nil); err != nil {
+		// For this testcase we don't want to enforce the FeeBuffer
```
nit: newline

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -10209,3 +10235,413 @@ func createRandomHTLC(t *testing.T, incoming bool) channeldb.HTLC {
+func TestApplyCommitmentFee(t *testing.T) {
+	var (
+		balance = lnwire.NewMSatFromSatoshis(5_000_000)
+		// balance used to test the case where the commitment fee
```
nit: newlines for all the comments

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -10209,3 +10235,413 @@ func createRandomHTLC(t *testing.T, incoming bool) channeldb.HTLC {
+		)
+		feeBuffer = CalcFeeBuffer(feePerKw, commitWeight)
+	)
+	// Create test channels so that we can use the `applyCommitFee`
```
nit: newline

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -10209,3 +10235,413 @@ func createRandomHTLC(t *testing.T, incoming bool) channeldb.HTLC {
+// TestApplyCommitmentFee tests that depending on the buffer type the correct
+// commitment fee is calculated which includes the buffer amount which will be
+// kept and is not usable hence not considered part of the usable local balance.
+func TestApplyCommitmentFee(t *testing.T) {
```
👍

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -10209,3 +10235,413 @@ func createRandomHTLC(t *testing.T, incoming bool) channeldb.HTLC {
+				return
+			}
+
+			require.NoError(t, err)
```
think we can just do `require.ErrorIs` here and no need to return early because there are returned values from the method even when there's an error.

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -10209,3 +10235,413 @@ func createRandomHTLC(t *testing.T, incoming bool) channeldb.HTLC {
+
+	// Create test channels to test out this state transition. The channel
+	// capactiy is 10 BTC with every side having 5 BTC at start. The fee
+	// rate is static and 6000 sats/kweight.
```
```suggestion
	// rate is static and 6000 sats/kw.
```

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -10209,3 +10235,413 @@ func createRandomHTLC(t *testing.T, incoming bool) channeldb.HTLC {
+
+	capacity := aliceChannel.channelState.Capacity
+
+	// Static fee rate of 6000 sats/kweigth.
```
```suggestion
	// Static fee rate of 6000 sats/kw.
```

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -10209,3 +10235,413 @@ func createRandomHTLC(t *testing.T, incoming bool) channeldb.HTLC {
+//	-----sig------>	|
+// 	(Alice fails with ErrBelowChanReserve)
+
+func TestAsynchronousSendingContraint(t *testing.T) {
```
Nice!

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -10209,3 +10235,413 @@ func createRandomHTLC(t *testing.T, incoming bool) channeldb.HTLC {
+
+	// Create test channels to test out this state transition. The channel
+	// capactiy is 10 BTC with every side having 5 BTC at start. The fee
+	// rate is static and 6000 sats/kweight.
```
```suggestion
	// rate is static and 6000 sats/kw.
```

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -10209,3 +10235,413 @@ func createRandomHTLC(t *testing.T, incoming bool) channeldb.HTLC {
+
+	capacity := aliceChannel.channelState.Capacity
+
+	// Static fee rate of 6000 sats/kweigth.
```
```suggestion
	// Static fee rate of 6000 sats/kw.
```

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -10209,3 +10235,413 @@ func createRandomHTLC(t *testing.T, incoming bool) channeldb.HTLC {
+	if err := ForceStateTransition(aliceChannel, bobChannel); err != nil {
+		t.Fatalf("unable to transition state: %v", err)
+	}
+	// Force a state transition, this will lock-in the htlc of bob.
```
nit: newline

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -10209,3 +10235,413 @@ func createRandomHTLC(t *testing.T, incoming bool) channeldb.HTLC {
+	// -----rev-----> (locks in alice's htlc)
+	// bob's htlc is still not fully locked in.
+	if err := ForceStateTransition(aliceChannel, bobChannel); err != nil {
+		t.Fatalf("unable to transition state: %v", err)
```
nit: use `require.NoError`

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -10209,3 +10235,413 @@ func createRandomHTLC(t *testing.T, incoming bool) channeldb.HTLC {
+	// <----rev------ (locks in bob's htlc)
+	// bob's htlc is still not fully locked in.
+	if err := ForceStateTransition(aliceChannel, bobChannel); err != nil {
+		t.Fatalf("unable to transition state: %v", err)
```
nit: use `require.NoError` and a few other places

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -10209,3 +10235,413 @@ func createRandomHTLC(t *testing.T, incoming bool) channeldb.HTLC {
+		t.Fatalf("unable to transition state: %v", err)
+	}
+
+	// Adding an HTLC from Alice Side should not be possible because
```
```suggestion
	// Adding an HTLC from Alice side should not be possible because
```

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -10209,3 +10235,413 @@ func createRandomHTLC(t *testing.T, incoming bool) channeldb.HTLC {
+	// -----sig-----> (includes alice htlc)
+	// <----rev------
+	// <----sig------ (includes alice and bobs htlc)
+	// -----rev-----> (locks in alice's htlc)
```
a bit confused here, you mean this locks Alice's htlc on Bob's local commitment?

ziggie1984 replied:
by locked in I mean that the HTLC is now present in both versions of the commitment transaction for both parties. Which means both parties have this htlc in their "tail" commitment transaction.

ziggie1984 replied:
but I think your wording is more on point so updated like so:

```locks in alice's htlc for bob```

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -10209,3 +10235,413 @@ func createRandomHTLC(t *testing.T, incoming bool) channeldb.HTLC {
+	// Now lets both channel parties know about these new htlcs.
+	// 	<----add-------	|
+	// 			|-----add------>
+	_, err = aliceChannel.ReceiveHTLC(htlc2)
```
I think this ADD will succeed because of the order? I think another good test case is, Bob sends an HTLC, Alice receives it, then Alice sends an HTLC, but it's failed due to the fee buffer limit.

ziggie1984 replied:
> I think this ADD will succeed because of the order?

Yes this ADD succeeds for Alice because she does not include her unacked HTLC yet, but this test case could also be constructed so that bob fails this Channel because of the BelowReserve problem.

ziggie1984 replied:
Added the new testcase

yyforyongyu commented on htlcswitch/link_test.go:
```diff
@@ -2927,22 +3017,22 @@ func TestChannelLinkTrimCircuitsNoCommit(t *testing.T) {
 	// Capture Alice's starting bandwidth to perform later, relative
 	// bandwidth assertions.
-	aliceStartingBandwidth := alice.link.Bandwidth()
+	// aliceStartingBandwidth := alice.link.Bandwidth()
```
typo?

yyforyongyu reviewed (changes requested):
Very nice👍 Think we can finally fix this issue, also great unit tests! No major comments, only questions and nits. Haven't finished `link_test.go` and itest yet, but I think this is veeeerry close!

ziggie1984 commented on lnwallet/channel.go:
```diff
@@ -3818,16 +3823,158 @@ func (lc *LightningChannel) getUnsignedAckedUpdates() []channeldb.LogUpdate {
+	commitWeight int64) lnwire.MilliSatoshi {
+
+	// Account for a 100% in fee rate increase.
+	bufferFeePerKw := 2 * feePerKw
```
Hmm, so I think we should always look at the current fee rate of the commitment transaction and go from there? As you correctly pointed out, these should be different for anchor and non-anchor channels (which currently they are but not in a smart way - for anchors its just hardcapped but should rather be at least floored at the min_relay fee of the avarage peers.) I agree that we should add a floor for the buffer in case we have very low feerates but I would do this in a future PR.

ziggie1984 commented on lnwallet/channel.go:
```diff
@@ -3887,6 +4004,51 @@ func (lc *LightningChannel) validateCommitmentSanity(theirLogCounter,
+	} else {
+		// No FeeBuffer is enforced when we are not the initiator of
+		// the channel. We cannot do this, because if our peer does not
+		// enforce the FeeBuffer (older LND software) the peer might
```
Yeah it will not fix 8315 until we carve-out the cases where we acutally allow the peer dip us below our reserve (outlined in the TODO).

> when it dips below fee buffer, fail the ADD but keep the link

I think this is already in place when this PR is merged, we do not account for the fee buffer when receiving, and when sending we can easily roll back the `ADD`.

ziggie1984 commented on lnwallet/channel.go:
```diff
@@ -5935,8 +6150,13 @@ func (lc *LightningChannel) ReceiveHTLC(htlc *lnwire.UpdateAddHTLC) (uint64, err
+	// We are not enforcing the FeeBuffer here because our peer is not
+	// aware of it and neither should they be. This is one of the reasons
+	// the FeeBuffer was introduced so we use it and therefore do not
+	// enforce the additional buffer.
```
Rephrased the comment, so the main reason we do not use it here is for the concurrency problem we try to solve:

```
	// We do not enforce the FeeBuffer here because one of the reasons it
	// was introduced is to protect against asynchronous sending of htlcs so
	// we use it here. The current lightning protocol does not allow to
	// reject ADDs already sent by the peer.
```

ziggie1984 replied:
The point you are making, that we should look at whether the htlc ADDs balance to our local will be implemented in a future PR, this relates to the "allow dipping the peer below the reserve" question.

ziggie1984 commented on lnwallet/channel.go:
```diff
@@ -4801,6 +4801,19 @@ func (lc *LightningChannel) computeView(view *htlcView, remoteChain bool,
+	// because MilliSathoshi is a unsigned type and can underflow in
+	// `evaluateHTLCView`. This should never happen for views which do not
+	// include new updates (remote or local).
+	if int64(ourBalance) < 0 {
```
I just moved this check to the root func. imo, was prior also in `validateCommitmentSanity`. This check is important to prevent the Peer mess with its balance from our point of view.

https://github.com/lightningnetwork/lnd/blob/master/lnwallet/channel.go#L3869-L3881

yyforyongyu commented on itest/lnd_multi-hop-error-propagation_test.go:
```diff
@@ -243,7 +243,8 @@ func testHtlcErrorPropagation(ht *lntest.HarnessTest) {
 	chanReserve := int64(chanAmt / 100)
-	amtToSend := int64(chanAmt) - chanReserve - 20000
+	feeBuffer := lntest.CalcStaticFeeBuffer(cType, 0)
+	amtToSend := int64(chanAmt) - chanReserve - int64(feeBuffer) - 10000
```
Are the above comments still correct?

ziggie1984 replied:
good catch, I think we can get rid of the for loop altogether.

yyforyongyu reviewed (approved):
LGTM⛵️ Thank you for all the effort, and, looking forward to the followup PR on fixing the dip-below-reserve-error on the remote side!

C-Otto commented on docs/release-notes/release-notes-0.18.0.md:
```diff
@@ -52,6 +52,15 @@
+* [Fixed](https://github.com/lightningnetwork/lnd/pull/8096) a case where `lnd`
+  might dip below its channel reserve when htlcs are added concurrently. A
+  fee buffer (additional balance) is now always kept on the local side ONLY
+  if the channel was opened locally. This is in accordance with the BOTL 02
```
nit: typo
